andela/ah-backend-athena
# #162163272 Users should be able to follow each other

#### What does this PR do?
This pull request allows authors to follow or unfollow each other. It also allows a user to view a list of their followers and a list of the authors they follow.

#### Description of tasks to be completed?

- write tests that fail
- add a Follow model to the `profiles` app
- add serializers to validate data
- add views to handle the different functions i.e. follow, unfollow, list all followers, list all followed
- add the views to paths in the URL patterns
- ensure that the tests are passing

#### How can this be manually tested?
- Use the link attached by Heroku i.e. deployed review app (unavailable for now) OR pull the branch to your local machine
- make sure you've created at least 2 users in the app. login as one of them.
- navigate to *POST* `/api/profiles/<username>/follow`. replace `<username>` with the other user's username. *DELETE* on the same endpoint will unfollow the user.
- it should return a profile showing that you are now following them
- navigating to *GET* `/api/profiles/following` shows the authors the logged in user is following.
- navigating to *GET* `/api/profiles/follower` shows the authors the logged in user is following.

#### What are the relevant pivotal stories?
#162163272

#### Screenshots
follow a user
![image](https://user-images.githubusercontent.com/5712103/49956130-883e9900-ff15-11e8-97e5-1a8b7ea86b8d.png)

unfollow a user
![image](https://user-images.githubusercontent.com/5712103/49956143-8bd22000-ff15-11e8-8b89-000b65f75e44.png)

trying to follow the same user
![image](https://user-images.githubusercontent.com/5712103/49956211-b1f7c000-ff15-11e8-9c42-d1f901441c59.png)

trying to follow yourself
![image](https://user-images.githubusercontent.com/5712103/49956329-fa16e280-ff15-11e8-9deb-15b09f448a1c.png)


show list of followed authors
![image](https://user-images.githubusercontent.com/5712103/49956426-31858f00-ff16-11e8-9761-61f50970fb45.png)

show list of users who follow the current user and if the current user follows them
![image](https://user-images.githubusercontent.com/5712103/49956485-57129880-ff16-11e8-95fc-8d8af320ed81.png)
